### ISSUE TITLE
Characterize Password Quality Using zxcvbn-python

### ISSUE THREAD

**Issue by joswr1ght** [clr2of8/DPAT#13]
Using [zxcvbn-python](https://github.com/dwolfhub/zxcvbn-python), characterize cracked passwords to report on the overall quality of passwords recovered. Zxcvbn provides easy access to detailed password strength indicators:

```
$ python -c "import pprint; from zxcvbn import zxcvbn; pp = pprint.PrettyPrinter(indent=4); pp.pprint(zxcvbn('Summer2020'))"
{   'calc_time': datetime.timedelta(microseconds=1474),
    'crack_times_display': {   'offline_fast_hashing_1e10_per_second': 'less '
                                                                       'than a '
                                                                       'second',
                               'offline_slow_hashing_1e4_per_second': '7 '
                                                                      'seconds',
                               'online_no_throttling_10_per_second': '2 hours',
                               'online_throttling_100_per_hour': '27 days'},
    'crack_times_seconds': {   'offline_fast_hashing_1e10_per_second': Decimal('0.0000065752'),
                               'offline_slow_hashing_1e4_per_second': Decimal('6.5752'),
                               'online_no_throttling_10_per_second': Decimal('6575.2'),
                               'online_throttling_100_per_hour': Decimal('2367072.000000000131398891767')},
    'feedback': {   'suggestions': [   'Add another word or two. Uncommon '
                                       'words are better.',
                                       "Capitalization doesn't help very "
                                       'much.'],
                    'warning': 'This is similar to a commonly used password.'},
    'guesses': Decimal('65752'),
    'guesses_log10': 4.817908967439416,
    'password': 'Summer2020',
    'score': 1,
    'sequence': [   {   'base_guesses': 69,
                        'dictionary_name': 'passwords',
                        'guesses': 138,
                        'guesses_log10': 2.1398790864012365,
                        'i': 0,
                        'j': 5,
                        'l33t': False,
                        'l33t_variations': 1,
                        'matched_word': 'summer',
                        'pattern': 'dictionary',
                        'rank': 69,
                        'reversed': False,
                        'token': 'Summer',
                        'uppercase_variations': 2},
                    {   'base_guesses': Decimal('101'),
                        'base_matches': [   {   'guesses': 100,
                                                'guesses_log10': 2.0,
                                                'i': 0,
                                                'j': 1,
                                                'pattern': 'bruteforce',
                                                'token': '20'}],
                        'base_token': '20',
                        'guesses': Decimal('202'),
                        'guesses_log10': 2.305351369446624,
                        'i': 6,
                        'j': 9,
                        'pattern': 'repeat',
                        'repeat_count': 2.0,
                        'token': '2020'}]}
```

The _score_ indicator would make for a useful reporting metric. Combined with group analysis, DPAT could order password selection by group by examining the average and mean score for passwords, even adding a little info-graphic of the score (like websites do when you enter a password to indicate quality).

I think it's important for DPAT to continue working without third-party modules. The addition of zxcvbn-python should be optional where `dpat.py` continues to operate, omitting the zxcvbn-python analysis features when the module is not installed (perhaps generating a warning to alert the user as such).

**Comment by JrOrOneEquals1** [comment]
@joswr1ght, I uploaded these changes to the password-strength branch.  Thanks for the input!

**Comment by joswr1ght** [comment]
I'm getting a div-by-zero error on my sample data:
[Wardrobe99-NTDS.zip](https://github.com/clr2of8/DPAT/files/4166663/Wardrobe99-NTDS.zip)

```
DPAT (password-strength) $ python dpat.py -n ../Wardrobe99-NTDS/Wardrobe99.ntds -c ~/.hashcat/hashcat.potfile  -g ../Wardrobe99-NTDS/groups/*
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Doesn't look like the Group Files are in the form output by PowerView, assuming the files are already in domain\username list form
Processing hash information from ../Wardrobe99-NTDS/Wardrobe99.ntds
Processing hash information from /Users/jwright/.hashcat/hashcat.potfile
Calculating password strength scores
Gathering hash info
Gathering group membership details
Getting group membership details and passwords cracked for AP Admin
Getting group membership details and passwords cracked for AP Staff
Getting group membership details and passwords cracked for AR Admin
Getting group membership details and passwords cracked for AR Staff
Getting group membership details and passwords cracked for Access Control Assistance Operators
Getting group membership details and passwords cracked for Account Operators
Getting group membership details and passwords cracked for Administrators
Getting group membership details and passwords cracked for Allowed RODC Password Replication Group
Getting group membership details and passwords cracked for Backup Operators
Getting group membership details and passwords cracked for CFO Admin
Getting group membership details and passwords cracked for CFO Staff
Getting group membership details and passwords cracked for Cert Publishers
Getting group membership details and passwords cracked for Certificate Service DCOM Access
Getting group membership details and passwords cracked for Cloneable Domain Controllers
Getting group membership details and passwords cracked for Cryptographic Operators
Getting group membership details and passwords cracked for Denied RODC Password Replication Group
Getting group membership details and passwords cracked for Distributed COM Users
Getting group membership details and passwords cracked for DnsAdmins
Getting group membership details and passwords cracked for DnsUpdateProxy
Getting group membership details and passwords cracked for Domain Admins
Getting group membership details and passwords cracked for Domain Computers
Getting group membership details and passwords cracked for Domain Controllers
Getting group membership details and passwords cracked for Domain Guests
Getting group membership details and passwords cracked for Domain Users
Getting group membership details and passwords cracked for Enterprise Admins
Getting group membership details and passwords cracked for Enterprise Key Admins
Getting group membership details and passwords cracked for Enterprise Read-only Domain Controllers
Getting group membership details and passwords cracked for Event Log Readers
Getting group membership details and passwords cracked for Finance
Getting group membership details and passwords cracked for Group Policy Creator Owners
Getting group membership details and passwords cracked for Guests
Getting group membership details and passwords cracked for HR Admin
Getting group membership details and passwords cracked for HR Staff
Getting group membership details and passwords cracked for Hyper-V Administrators
Getting group membership details and passwords cracked for IIS_IUSRS
Getting group membership details and passwords cracked for IT Admin
Getting group membership details and passwords cracked for IT Staff
Getting group membership details and passwords cracked for Incoming Forest Trust Builders
Getting group membership details and passwords cracked for Key Admins
Getting group membership details and passwords cracked for LLG Admin
Getting group membership details and passwords cracked for LLG Staff
Getting group membership details and passwords cracked for Network Configuration Operators
Getting group membership details and passwords cracked for OPS
Getting group membership details and passwords cracked for Performance Log Users
Getting group membership details and passwords cracked for Performance Monitor Users
Getting group membership details and passwords cracked for Pre-Windows 2000 Compatible Access
Getting group membership details and passwords cracked for Print Operators
Getting group membership details and passwords cracked for Protected Users
Getting group membership details and passwords cracked for RAS and IAS Servers
Getting group membership details and passwords cracked for RDS Endpoint Servers
Getting group membership details and passwords cracked for RDS Management Servers
Getting group membership details and passwords cracked for RDS Remote Access Servers
Getting group membership details and passwords cracked for Read-only Domain Controllers
Getting group membership details and passwords cracked for Remote Desktop Users
Getting group membership details and passwords cracked for Remote Management Users
Getting group membership details and passwords cracked for Replicator
Getting group membership details and passwords cracked for SRC Admin
Getting group membership details and passwords cracked for SRC Staff
Getting group membership details and passwords cracked for Schema Admins
Getting group membership details and passwords cracked for Server Operators
Getting group membership details and passwords cracked for Storage Replica Administrators
Getting group membership details and passwords cracked for Terminal Server License Servers
Getting group membership details and passwords cracked for Users
Getting group membership details and passwords cracked for W99-XS1
#!/usr/bin/python
from __future__ import print_function
try:
    input = raw_input
except NameError:
    pass
import webbrowser, io, os, re, argparse, sqlite3, sys, time, binascii, hashlib
pwStren = True
try:
    from zxcvbn import zxcvbn
except ImportError:
    pwStren = False
from shutil import copyfile
try:
:528
Getting group membership details and passwords cracked for Windows Authorization Access Group
Gathering password length statistics
Gathering password reuse statistics
Gathering password history details
Traceback (most recent call last):
  File "dpat.py", line 528, in <module>
    list[list.index(item)] = [item, str(round(strengths[0][1]/float(amount) * 100, 1)) + '%', str(round(strengths[1][1]/float(amount) * 100, 1)) + '%', str(round(strengths[2][1]/float(amount) * 100, 1)) + '%', str(round(strengths[3][1]/float(amount) * 100, 1)) + '%', str(round(strengths[4][1]/float(amount) * 100, 1)) + '%', str(round(total/float(amount) * 25))]
ZeroDivisionError: float division by zero
```


Sample data attached. Here is the `secretsdump.py` and `hashcat` syntax I used:

```
$ secretsdump.py -history -system registry/SYSTEM -ntds Active\ Directory/ntds.dit LOCAL -outputfile Wardrobe99
$ hashcat -a 0 -m 1000 Wardrobe99.ntds ~/Hack/rockyou.txt
$ cd ../DPAT/
$ python dpat.py -n ../Wardrobe99-NTDS/Wardrobe99.ntds -c ~/.hashcat/hashcat.potfile -history  -g ../Wardrobe99-NTDS/groups/*
```

**Comment by JrOrOneEquals1** [comment]
Thank you for taking a look. I see the issue and will work on it. (I'm out for the next week and will have to look at it the week after next).

**Comment by joswr1ght** [comment]
Thanks! I'm not in a rush. I'm finishing up a year-long project soon, and I hope to have time to contribute some code as well.

**Comment by clr2of8** [comment]
Divide by zero error fixed in PR #20 by @JrOrOneEquals1. Feature is complete now. Thank you!
